File: platform/watchos/src/iosMain/swift/MeBeatMeViewModel.swift
import Foundation
import HealthKit
import WorkoutKit
import Combine
import core

class MeBeatMeViewModel: ObservableObject {
    @Published var currentScreen: Screen = .challengeSelection
    @Published var challenges: [BeatChoice] = []
    @Published var liveSession: LiveSession?
    @Published var isOnTarget: Bool = false
    @Published var lastScore: Score?
    
    private let healthStore = HKHealthStore()
    private let workoutManager = WorkoutManager()
    private let historyStore = HistoryStore()
    private var cancellables = Set<AnyCancellable>()
    
    func requestHealthPermissions() {
        let typesToRead: Set<HKObjectType> = [
            HKObjectType.quantityType(forIdentifier: .distanceWalkingRunning)!,
            HKObjectType.quantityType(forIdentifier: .activeEnergyBurned)!,
            HKObjectType.workoutType()
        ]
        
        healthStore.requestAuthorization(toShare: nil, read: typesToRead) { success, error in
            if let error = error {
                print("Health authorization error: \(error)")
            }
        }
    }
    
    func generateChallenges() {
        let bucket = Bucket.km_3_8
        let planner = BeatPlanner(history: historyStore.all())
        challenges = planner.choicesFor(bucket: bucket)
    }
    
    func startLiveRun(_ choice: BeatChoice) {
        let session = LiveSession(
            choice: choice,
            startTimeMs: Int64(Date().timeIntervalSince1970 * 1000)
        )
        liveSession = session
        currentScreen = .liveRun
        
        // Start workout tracking
        workoutManager.startWorkout { [weak self] workoutData in
            DispatchQueue.main.async {
                self?.updateLiveSession(workoutData)
            }
        }
    }
    
    private func updateLiveSession(_ workoutData: WorkoutData) {
        guard var session = liveSession else { return }
        
        session = session.copy(
            currentDistanceM: workoutData.distanceMeters,
            currentElapsedSec: workoutData.elapsedSeconds,
            currentPaceSecPerKm: workoutData.currentPaceSecPerKm
        )
        
        liveSession = session
        
        // Check if on target pace
        isOnTarget = session.isOnTargetPace()
        
        // Trigger haptic feedback
        if isOnTarget {
            WKInterfaceDevice.current().play(.success)
        }
    }
    
    func stopLiveRun() {
        guard let session = liveSession else { return }
        
        workoutManager.stopWorkout()
        
        // Complete the session
        let runSession = session.complete()
        let ppi = PpiCurve.score(distanceM: runSession.distanceMeters, elapsedSec: runSession.elapsedSeconds)
        let bucket = bucketFor(distanceM: runSession.distanceMeters)
        
        let score = Score(
            sessionId: runSession.id,
            ppi: ppi,
            bucket: bucket
        )
        
        // Save to history
        historyStore.add(score: score)
        
        lastScore = score
        liveSession = nil
        currentScreen = .postRun
        
        // Celebration haptic
        WKInterfaceDevice.current().play(.success)
    }
    
    func startNewSession() {
        currentScreen = .challengeSelection
        lastScore = nil
        isOnTarget = false
        generateChallenges()
    }
}

// WorkoutManager for HealthKit integration
class WorkoutManager: NSObject, ObservableObject {
    private let healthStore = HKHealthStore()
    private var workoutSession: HKWorkoutSession?
    private var workoutBuilder: HKLiveWorkoutBuilder?
    
    func startWorkout(completion: @escaping (WorkoutData) -> Void) {
        let configuration = HKWorkoutConfiguration()
        configuration.activityType = .running
        configuration.locationType = .outdoor
        
        do {
            workoutSession = try HKWorkoutSession(healthStore: healthStore, configuration: configuration)
            workoutBuilder = workoutSession?.associatedWorkoutBuilder()
            
            workoutBuilder?.dataSource = HKLiveWorkoutDataSource(healthStore: healthStore, workoutConfiguration: configuration)
            
            workoutSession?.startActivity(with: Date())
            workoutBuilder?.beginCollection(at: Date()) { success, error in
                if let error = error {
                    print("Workout builder error: \(error)")
                }
            }
            
            // Monitor workout data
            workoutBuilder?.dataSource?.enableCollection(for: HKQuantityType.quantityType(forIdentifier: .distanceWalkingRunning)!, predicate: nil)
            workoutBuilder?.dataSource?.enableCollection(for: HKQuantityType.quantityType(forIdentifier: .activeEnergyBurned)!, predicate: nil)
            
            // Start monitoring
            Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { _ in
                self.updateWorkoutData(completion: completion)
            }
            
        } catch {
            print("Workout session error: \(error)")
        }
    }
    
    private func updateWorkoutData(completion: @escaping (WorkoutData) -> Void) {
        guard let builder = workoutBuilder else { return }
        
        let distance = builder.statistics(for: HKQuantityType.quantityType(forIdentifier: .distanceWalkingRunning)!)?.sumQuantity()?.doubleValue(for: HKUnit.meter()) ?? 0.0
        let duration = builder.elapsedTime
        let pace = duration > 0 ? (duration / (distance / 1000.0)) : 0.0
        
        let workoutData = WorkoutData(
            distanceMeters: distance,
            elapsedSeconds: duration,
            currentPaceSecPerKm: pace
        )
        
        completion(workoutData)
    }
    
    func stopWorkout() {
        workoutSession?.end()
        workoutBuilder?.endCollection(at: Date()) { success, error in
            if let error = error {
                print("End collection error: \(error)")
            }
        }
    }
}

struct WorkoutData {
    let distanceMeters: Double
    let elapsedSeconds: Double
    let currentPaceSecPerKm: Double
}

File: platform/wearos/src/main/kotlin/MeBeatMeViewModel.kt
package com.mebeatme.wearos

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mebeatme.core.ppi.PpiEngine
import com.mebeatme.wearos.data.ScoreDao
import com.mebeatme.wearos.data.ScoreEntity
import com.mebeatme.wearos.health.HealthServicesManager
import com.mebeatme.wearos.health.WorkoutData
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import android.content.Context

class MeBeatMeViewModel(private val scoreDao: ScoreDao) : ViewModel() {
    
    private val _currentScreen = MutableStateFlow(Screen.ChallengeSelection)
    val currentScreen: StateFlow<Screen> = _currentScreen.asStateFlow()
    
    private val _liveSession = MutableStateFlow<LiveSession?>(null)
    val liveSession: StateFlow<LiveSession?> = _liveSession.asStateFlow()
    
    private val _isOnTarget = MutableStateFlow(false)
    val isOnTarget: StateFlow<Boolean> = _isOnTarget.asStateFlow()
    
    private val _lastScore = MutableStateFlow<Score?>(null)
    val lastScore: StateFlow<Score?> = _lastScore.asStateFlow()
    
    private var healthServicesManager: HealthServicesManager? = null
    
    fun initializeHealthServices(context: Context) {
        healthServicesManager = HealthServicesManager(context)
    }
    
    fun startLiveRun(choice: BeatChoice) {
        val session = LiveSession(
            choice = choice,
            startTimeMs = System.currentTimeMillis(),
            targetPPI = getTargetPPI()
        )
        _liveSession.value = session
        _currentScreen.value = Screen.LiveRun
        
        // Start health services workout
        viewModelScope.launch {
            healthServicesManager?.startWorkout()?.collect { workoutData ->
                updateLiveSession(workoutData)
            }
        }
    }
    
    private fun updateLiveSession(workoutData: WorkoutData) {
        val currentSession = _liveSession.value ?: return
        
        val updatedSession = currentSession.copy(
            currentDistanceM = workoutData.distanceMeters,
            currentElapsedSec = workoutData.elapsedSeconds,
            currentPaceSecPerKm = workoutData.currentPaceSecPerKm,
            targetPPI = currentSession.targetPPI
        )
        
        _liveSession.value = updatedSession
        
        // Check if on target pace
        val onTarget = updatedSession.isOnTargetPace()
        _isOnTarget.value = onTarget
    }
    
    fun stopLiveRun() {
        val session = _liveSession.value ?: return
        
        viewModelScope.launch {
            // Stop health services
            healthServicesManager?.stopWorkout()
            
            // Complete the session
            val runSession = session.complete()
            val ppi = PpiEngine.score(runSession.distanceMeters, runSession.elapsedSeconds)
            val bucket = bucketFor(runSession.distanceMeters)
            
            val score = Score(
                sessionId = runSession.id,
                ppi = ppi,
                bucket = bucket
            )
            
            // Save to database
            scoreDao.insertScore(ScoreEntity.fromScore(score))
            
            _lastScore.value = score
            _liveSession.value = null
            _currentScreen.value = Screen.PostRun
        }
    }
    
    fun startNewSession() {
        _currentScreen.value = Screen.ChallengeSelection
        _lastScore.value = null
        _isOnTarget.value = false
    }
    
    fun getTargetPPI(): Double {
        // Get the highest PPI from the last 90 days
        // This would typically come from a database or shared service
        // For now, we'll use a placeholder value
        return 300.0 // This should be injected from a proper data source
    }
}
